web/client-api/src/main/java/io/deephaven/web/client/api/input/JsInputTable.java

```diff
@@ -60,99 +64,155 @@ public JsInputTable(JsTable from, String[] keys, String[] values) {
-
-                rowValues[i] = new RowValues(v.toArray(new ColumnValue[v.size()]));
+    public Promise<JsInputTable> addRows(JsPropertyMap<?>[] rows, @JsOptional String userTimeZone) {
+        String[] names =
+                Arrays.stream(table.lastVisibleState().getColumns()).map(Column::getName).toArray(String[]::new);
+        String[] types =
+                Arrays.stream(table.lastVisibleState().getColumns()).map(Column::getType).toArray(String[]::new);
+
+        Object[][] data = new Object[names.length][];
+        for (int i = 0; i < names.length; i++) {
+            String name = names[i];
+            Object[] columnArray = new Object[rows.length];
+            for (int j = 0; j < rows.length; j++) {
+                columnArray[j] = rows[j].get(name);
             }
-            // table.getServer().addRowsToInputTable(table.getHeadHandle(), rowValues, c);
-            throw new UnsupportedOperationException("addRowsToInputTable");
-        }).then(response -> Promise.resolve(this));
+            data[i] = columnArray;
+        }
+
+        return table.getConnection().newTable(names, types, data, userTimeZone, null)
+                .then(this::addTable);
     }
 
     public Promise<JsInputTable> addTable(JsTable tableToAdd) {
         return addTables(new JsTable[] {tableToAdd});
     }
 
     public Promise<JsInputTable> addTables(JsTable[] tablesToAdd) {
-        return Callbacks.<Void, String>promise(this.table, c -> {
-            // table.getServer().addTablesToInputTable(table.getHeadHandle(), Arrays.stream(tablesToAdd).map(t ->
-            // t.getHandle()).toArray(TableHandle[]::new), c);
-            throw new UnsupportedOperationException("addTablesToInputTable");
-        }).then(response -> Promise.resolve(this));
+        if (tablesToAdd.length == 0) {
+            // noinspection unchecked,rawtypes
+            return (Promise) Promise.reject("Must provide at least one table");
+        }
+        final Promise<JsTable> mergePromise;
+        if (tablesToAdd.length == 1) {
+            mergePromise = Promise.resolve(tablesToAdd[0]);
+        } else {
+            mergePromise = table.getConnection().mergeTables(tablesToAdd, this.table);
+        }
+
+        return mergePromise
+                .then(merged -> {
+                    // noinspection CodeBlock2Expr - easier readability for chained then()
+                    return Callbacks.grpcUnaryPromise(c -> {
+                        AddTableRequest addTableRequest = new AddTableRequest();
+                        addTableRequest.setInputTable(table.getHeadHandle().makeTicket());
+                        addTableRequest.setTableToAdd(merged.getHeadHandle().makeTicket());
+                        table.getConnection().inputTableServiceClient().addTableToInputTable(addTableRequest,
+                                table.getConnection().metadata(), c::apply);
+                    }).then(success -> {
+                        if (merged != tablesToAdd[0]) {
```

Comment:
I understand the check; but correctness might be more obvious if you stick to the length check like on line 101.

Reply:
The goal in having a different check was to acknowledge "this is an intermediate instance so it must be closed" - if the logic between here and 101 changes, I would want to allow for this still to be correct.

What do you think of something instead like `Arrays.stream(tablesToAdd).allMatch(t -> t != merged)`?